specs/AppLifecycle/Instancing/AppLifecycle SingleMulti-Instancing.md

```diff
@@ -0,0 +1,422 @@
+    }
+    else if (kind == ExtendedActivationKind::Protocol)
+    {
+        auto protocolArgs = args.Data().as<ProtocolActivatedEventArgs>();
+        DoSomethingWithProtocolArgs(protocolArgs.Uri());
+    }
+    else if (kind == ExtendedActivationKind::StartupTask)
+    {
+        auto startupArgs = args.Data().as<StartupTaskActivatedEventArgs>();
+        DoSomethingWithStartupArgs(startupArgs.TaskId());
+    }
+}
+```
+
+### Redirection orchestration
+
+We can also imagine a more advanced scenario, where the app registers an instance for redirection,
+but when it receives a redirected activation, it may apply further app logic to decide whether to
+handle this new activation in the current instance or redirect again, effectively chaining
+redirections.
+
+```c++
+void OnActivated(ActivationArguments const& args)
+{
+    const ExtendedActivationKind kind = args.Kind();
+
+    // For example, we might want to redirect protocol activations.
+    if (kind == ExtendedActivationKind::Protocol)
+    {
+        auto protocolArgs = args.Data().as<ProtocolActivatedEventArgs>();
+        Uri uri = protocolArgs.Uri();
+
+        // We'll try to find the instance that handles protocol activations.
+        // If there isn't one, then this instance will take over that duty.
+        auto instance = AppInstance::FindOrRegisterForKey(uri.AbsoluteUri().c_str());
+        if (!instance.IsCurrent())
+        {
+            instance.RedirectTo(args);
+        }
+        else
+        {
+            DoSomethingWithProtocolArgs(uri);
+        }
+    }
+    else
+    {
+        // In this example, this instance of the app handles all other
+        // activation kinds.
+        DoSomethingWithNewActivationArgs(args);
+    }
+}
+```
+
+In the existing UWP API, RedirectActivationTo does not take any parameters. However, in the UWP
+case, the activation contract is tightly controlled, and we could ensure that the arguments passed
+on to the app were the arguments for the redirected instance. In bringing this feature to Win32, we
+can't control activation as tightly. Also, the ActivationArguments object returned by
+GetActivatedEventArgs is not the same object as the IActivatedEventArgs object that is passed in to
+the Activated event callback. The new API gives the app the opportunity to modify or replace the
+payload to be passed to the target instance.
+
+### Unregister for redirection
+
+Just as an app can define its own logic for how and when it registers an instance for redirection,
+so too will it be able to define its own logic for when a given instance no longer wants to be a
+target for redirection.
+
+Following the previous example of an app that supports filetype activation, and redirects to the
+instance that already has the selected file open - when that file is closed, the app could
+unregister itself.
+
+```c++
+void CALLBACK OnFileClosed(HWND hDlg, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    AppInstance::GetCurrent().UnregisterKey(g_fileName);
```

Comment:
What happens if the app exits without unregistering the key, and as it's on its way out, somebody tries to redirect to it? Does the redirection silently disappear?

How long should an app linger after unregistering the key to make sure that any pending redirections have been drained?

Reply:
DISCUSSION: The lifetime of the registration will be pulled out on process termination, but an app needs to proactively remove registrations during shutdown. The UWP activation management correctly determines whether the target instance had terminated, and spin up a new instance to handle the lingering redirection.

CONSIDER: Figure out a way to signal back to the original redirector. Might imply a change to the `Redirect...` call, including some form of `Async` behavior so apps don't have to block.  There's a race between apps who want to redirect to another app and the target deciding to terminate.

Reply:
Added further description.